### DISCUSSION TITLE
Peer review request and best-practice production-ready code repositories request

### DISCUSSION THREAD

**Discussion by LucaMaurelliC** [sqlalchemy/sqlalchemy#10484] [Usage Questions]
Hi,

TD:DR: I'm new to SQL Alchemy and databases in general, and recently I commited myself to refactor some very old business logic code written with no knowledge so far. Here I will write where I start and what I think I need to refactor the code with the ideas I read so far. My request is to have a peer review of this ideas with users of SQL Alchemy that may share their experience and guide me. Also I would like to ask if there are production-ready code repositories that use SQL Alchemy Core so that I can feel how code is handled.

To date I start with:
1. Queries are written as f-formatted strings and if a list of parameters is given, the logic to execute the query on the list of parameters is hard-code by parsing the f-formatted string. What I want to do is convert from f-formatting to named-parameters queries and leaving the comcumber logic of parsing the parameters to sql alchemy, e.g. by exploiting .execute() with a list of parameters (that should call executemany under the hood AFAIK).
2. Queries and connections were managed, with poor and sometimes wrong result, at low-level by database drivers/cursors. The aim is to use SQL Alchemy and leave the management of connection pools to it. Later I would like to switch to async.
3. My knowledge of how should I exploit the library in order to manage how transaction work and how should they be managed is limited. In other words, I found myself with three main uses cases at the moment:
A) A single "READ" query is performed, the result needs to be fetched and parsed.
B) A single "WRITE" query is performed, there is no result but the transaction needs to be commited
C) Multiple "WRITE" queries are performed, there is no result but the transactions needs to be commited (e.g. at the end of all queries I commit).
To date, the logic code get the query as a string and its type (read or write) and then execute it. Thinking about the refactoring with SQL Alchemy is there something I should be thinking with these three uses cases? Can I simplify things? Can I be more flexible and efficient?
4) What if I have two databases engines, and I run two writing queries respectively on the two engines. Can I have atomicity here so that they are commited only if both engines are ready to commit it?

Given that basically I'm new and I have to refactor/rewrite core business logic code from almost stratch, can you comment on these ideas or guide me where to look at? I also would like to read production-ready repository code that exploits SQL Alchemy in order to feel how things should be handled. Also, given where I start, I believe CORE is preferred in my use case, given also I don't like ORM. I also read the documention here and there already.

**Comment by zzzeek** [comment]
transaction control depends a lot on what kind of application this is (like a web app, or not) and how operations are organized.   web apps have the most straightforward transaction control model whereas something like a GUI app where the user clicks buttons and DB changes occur is probably the most complicated.

overall when using SQLAlchemy as well as the raw DBAPI in any case, the transaction control model is based on a non-autocommit model, where you have a database connection and whatever happens on it is not committed until you call commit().       pep-249 was written before Python had context managers or even decorators, but most DBAPIs have context manager patterns available, as does SQLAlchemy.

The context manager patterns you see in the tutorial at https://docs.sqlalchemy.org/en/20/tutorial/dbapi_transactions.html#committing-changes are the typical places you'd start.    at the same time you might want to take care to be calling a new transaction at the correct level, the classic example is a web application that within the span of one request will do three read queries and four write queries; usually you'd want all those things to happen in just one transaction, rather than seven tiny transactions, as web requests are most naturally scoped as transaction-per-request.      so I think the main thing to consider is trying to pull up the transaction to be as coarse grained as is appropriate (that is, have more operations in one enclosing transaction rather than a lot of little transactions).   

Generally I usually want business logic itself to not really be aware of transactions at all, it all assumes it's occurring within an enclosing scope that handles the lifespan of a tranasction (this is an old habit I got back from working with J2EE in the 1990s but it still holds up today).

for item 4 there is a technology called [2pc](https://en.wikipedia.org/wiki/Two-phase_commit_protocol) that's the "industry standard" way to get two separate database transactions to coordinate with each other atomically.   It's really not well supported by Python DBAPIs though, PostgreSQL (maybe SQL Server or Oracle) is the only platform where we even have 2pc working at all and it's very difficult to work with - the transaction that's created outlives your actual application if you dont close it cleanly, and there are other driver bumps and issues that exist here.  SQLAlchemy has a little bit of support for 2pc through the ORM session which you can see here: https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#session-twophase  but this is not stuff people normally use.   you can try working w/ 2pc but I'd keep it at arm's length since it might be difficult to keep running smoothly in production.
